File: src/main/java/net/dashflight/data/helpers/caching/Computable.java
package net.dashflight.data.helpers.caching;

/**
 * Computes a result based on an input
 * @param <K> The input type
 * @param <V> The result type
 */
public interface Computable<K, V> {

    V compute(K key) throws DataFetchException, InterruptedException;


    /**
     * Thrown when fetching a result fails.
     */
    class DataFetchException extends Exception {

        public DataFetchException(String message) {
            super(message);
        }

    }
}

File: src/main/java/net/dashflight/data/caching/Computable.kt
package net.dashflight.data.caching

/**
 * Computes a result based on an input
 * @param <K> The input type
 * @param <V> The result type
</V></K> */
interface Computable<K, V> {

    @Throws(DataFetchException::class, InterruptedException::class)
    fun compute(key: K): V

    /**
     * Thrown when fetching a result fails.
     */
    class DataFetchException(message: String) : Exception(message)
}